File: src/llm/ollama-client.ts
import { Ollama } from 'ollama';
import { A11yIssue } from '../state/types';

interface AnalysisResult {
  issues: A11yIssue[];
  summary: string;
}

export class OllamaClient {
  private ollama: Ollama;
  private model: string;

  constructor(host: string = 'http://localhost:11434', model: string = 'qwen2.5-coder:32b', apiKey?: string) {
    this.model = model;
    
    const config: { host: string; headers?: Record<string, string> } = {
      host: host.replace(/\/$/, ''),
    };

    if (apiKey) {
      config.headers = {
        Authorization: `Bearer ${apiKey}`,
      };
    }

    this.ollama = new Ollama(config);
  }

  async analyze(diffContent: string, prompt: string): Promise<AnalysisResult> {
    const fullPrompt = `${prompt}\n\n---\n\n## Code Diff:\n\n${diffContent}\n\nRespond with valid JSON only.`;

    try {
      const response = await this.ollama.chat({
        model: this.model,
        messages: [
          { role: 'user', content: fullPrompt }
        ],
        format: 'json',
        options: {
          temperature: 0.1,
          num_ctx: 32768,
        },
      });

      const content = response.message?.content || '';
      
      if (!content) {
        throw new Error('Empty response from Ollama');
      }

      return this.parseResponse(content);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      
      if (message.includes('401') || message.includes('Unauthorized')) {
        throw new Error(
          `Ollama authentication failed. For Ollama Cloud, ensure:\n` +
          `  1. You have a valid API key from https://ollama.com/settings/keys\n` +
          `  2. Set the 'api-key' input or OLLAMA_API_KEY environment variable\n` +
          `  3. Use 'ollama-url: https://ollama.com'\n` +
          `Original error: ${message}`
        );
      }
      
      throw new Error(`Ollama API error: ${message}`);
    }
  }

  private parseResponse(content: string): AnalysisResult {
    try {
      // Try parsing the entire content as JSON first
      try {
        const parsed = JSON.parse(content);
        return this.extractIssues(parsed);
      } catch {
        // Content might have extra text, extract JSON object
      }

      // Find JSON object by matching brackets
      const startIndex = content.indexOf('{');
      if (startIndex === -1) {
        throw new Error('No JSON object found in response');
      }

      let depth = 0;
      let endIndex = startIndex;

      for (let i = startIndex; i < content.length; i++) {
        if (content[i] === '{') depth++;
        if (content[i] === '}') depth--;
        if (depth === 0) {
          endIndex = i + 1;
          break;
        }
      }

      const jsonStr = content.substring(startIndex, endIndex);
      const parsed = JSON.parse(jsonStr);
      return this.extractIssues(parsed);
    } catch (parseError) {
      console.error('Raw Ollama response:', content.substring(0, 500));
      throw new Error(`Failed to parse Ollama response: ${parseError instanceof Error ? parseError.message : String(parseError)}`);
    }
  }

  private extractIssues(parsed: Record<string, unknown>): AnalysisResult {
    const issues: A11yIssue[] = ((parsed.issues as Array<Record<string, unknown>>) || []).map((issue) => {
      const rawSeverity = String(issue.severity || 'GOOD_PRACTICE').toUpperCase();
      const severity: A11yIssue['severity'] = rawSeverity === 'VIOLATION' ? 'VIOLATION' : 'GOOD_PRACTICE';
      
      return {
        file: String(issue.file || ''),
        line: issue.line ? Number(issue.line) : null,
        wcag_criterion: String(issue.wcag_criterion || ''),
        wcag_level: String(issue.wcag_level || 'A'),
        severity,
        title: String(issue.title || ''),
        description: String(issue.description || ''),
        suggestion: String(issue.suggestion || ''),
      };
    });

    return {
      issues,
      summary: String(parsed.summary || 'Accessibility review completed.'),
    };
  }
}

File: src/llm/ollama_client.py
"""Ollama API client for accessibility review."""

import json
import requests
from typing import Dict, List, Optional, Any

from .base import LLMClient, LLMResponse


class OllamaClient(LLMClient):
    """Ollama API client implementation."""

    def __init__(
        self,
        api_url: str = 'http://localhost:11434',
        model: str = 'qwen2.5-coder:32b'
    ):
        """
        Initialize the Ollama client.

        Args:
            api_url: Ollama API base URL
            model: Model name to use
        """
        self._api_url = api_url.rstrip('/')
        self._model_name = model
        self._timeout = 300  # 5 minutes timeout for long analyses

    @property
    def model_name(self) -> str:
        return self._model_name

    @property
    def backend_type(self) -> str:
        return 'ollama'

    def _make_request(
        self,
        endpoint: str,
        data: Dict,
        stream: bool = False
    ) -> Dict:
        """Make a request to the Ollama API."""
        url = f"{self._api_url}/api/{endpoint}"

        try:
            response = requests.post(
                url,
                json=data,
                timeout=self._timeout,
                headers={'Content-Type': 'application/json'}
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            raise RuntimeError(f"Ollama API error: {str(e)}")

    def _ensure_model_available(self):
        """Pull the model if not available locally."""
        try:
            # Check if model exists
            response = requests.get(f"{self._api_url}/api/tags", timeout=30)
            response.raise_for_status()
            models = response.json().get('models', [])

            model_names = [m.get('name', '').split(':')[0] for m in models]

            # Check if model or base model name exists
            base_model = self._model_name.split(':')[0]
            if self._model_name not in [m.get('name') for m in models] and base_model not in model_names:
                print(f"Model {self._model_name} not found, pulling...")
                self._pull_model()
        except Exception as e:
            raise RuntimeError(f"Failed to check/pull model: {str(e)}")

    def _pull_model(self):
        """Pull the model from Ollama registry."""
        response = requests.post(
            f"{self._api_url}/api/pull",
            json={'name': self._model_name, 'stream': False},
            timeout=600  # 10 minutes for model pull
        )
        response.raise_for_status()

    def analyze_diff(
        self,
        diff_content: str,
        system_prompt: str,
        user_prompt: str,
        json_schema: Optional[Dict] = None
    ) -> LLMResponse:
        """Analyze diff using Ollama API."""
        # Ensure model is available
        self._ensure_model_available()

        # Build the request
        full_prompt = f"{system_prompt}\n\n{user_prompt}\n\n## Diff to Analyze:\n\n{diff_content}"

        request_data = {
            'model': self._model_name,
            'prompt': full_prompt,
            'stream': False,
            'options': {
                'temperature': 0.1,
                'top_p': 0.95,
                'num_ctx': 32768,  # Large context for diffs
            }
        }

        # Add structured output format if schema provided
        if json_schema:
            request_data['format'] = 'json'
            # Add schema hint in the prompt
            request_data['prompt'] = (
                f"{full_prompt}\n\n"
                f"IMPORTANT: Respond with valid JSON matching this schema:\n"
                f"{json.dumps(json_schema, indent=2)}\n\n"
                f"Respond ONLY with valid JSON, no markdown formatting."
            )

        try:
            response = self._make_request('generate', request_data)

            content = response.get('response', '')

            # Parse structured data if JSON schema was requested
            structured_data = None
            if json_schema and content:
                try:
                    parsed = json.loads(content)
                    if isinstance(parsed, list):
                        structured_data = parsed
                    elif isinstance(parsed, dict) and 'issues' in parsed:
                        structured_data = parsed['issues']
                except json.JSONDecodeError:
                    # Try to extract JSON from response
                    content = self._extract_json(content)
                    if content:
                        try:
                            parsed = json.loads(content)
                            if isinstance(parsed, list):
                                structured_data = parsed
                            elif isinstance(parsed, dict) and 'issues' in parsed:
                                structured_data = parsed['issues']
                        except json.JSONDecodeError:
                            pass

            # Extract usage info
            usage = None
            if 'prompt_eval_count' in response or 'eval_count' in response:
                usage = {
                    'prompt_tokens': response.get('prompt_eval_count', 0),
                    'completion_tokens': response.get('eval_count', 0),
                    'total_tokens': response.get('prompt_eval_count', 0) + response.get('eval_count', 0)
                }

            return LLMResponse(
                content=content,
                model=self._model_name,
                usage=usage,
                structured_data=structured_data,
                raw_response=response
            )

        except Exception as e:
            raise RuntimeError(f"Ollama generation error: {str(e)}")

    def _extract_json(self, content: str) -> Optional[str]:
        """Extract JSON from content that might have surrounding text."""
        # Try to find JSON array or object
        import re

        # Try to match JSON array
        array_match = re.search(r'\[[\s\S]*?\]', content)
        if array_match:
            return array_match.group(0)

        # Try to match JSON object
        object_match = re.search(r'\{[\s\S]*?\}', content)
        if object_match:
            return object_match.group(0)

        return None

    def health_check(self) -> bool:
        """Check if Ollama server is running and model is available."""
        try:
            response = requests.get(f"{self._api_url}/api/tags", timeout=10)
            return response.status_code == 200
        except Exception:
            return False

    def get_model_info(self) -> Dict[str, Any]:
        """Get information about the Ollama model."""
        info = {
            'backend': 'ollama',
            'model': self._model_name,
            'api_url': self._api_url,
            'provider': 'Local',
            'supports_json_schema': True
        }

        # Try to get model details
        try:
            response = requests.post(
                f"{self._api_url}/api/show",
                json={'name': self._model_name},
                timeout=10
            )
            if response.status_code == 200:
                model_info = response.json()
                info['model_size'] = model_info.get('size')
                info['model_family'] = model_info.get('details', {}).get('family')
        except Exception:
            pass

        return info